File: software/src/main/java/com/example/interceptor/LoginInterceptor.java
package com.example.interceptor;


import java.io.PrintWriter;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;

/**
 * 登录拦截器（判断用户是否登录）
 */
public class LoginInterceptor extends HandlerInterceptorAdapter {

    @Override
    public boolean preHandle(HttpServletRequest request,
                             HttpServletResponse response, Object handler) throws Exception {
        response.setCharacterEncoding("utf-8");
        response.setContentType("text/html;charset=utf-8");
        // 从session中获取用户对象（该对象在用户登录的时候存于session中）
        Object object = request.getSession().getAttribute("user");
        // 用户未登录，则返回false, 拦截该请求
        if (object == null) {
            PrintWriter printWriter = response.getWriter();
            printWriter.write("请先登录");
            return false;
        }
        // 用户已登录，则返回true, 放行该请求
        return true;
    }

    @Override
    public void afterCompletion(HttpServletRequest request,
                                HttpServletResponse response, Object handler, Exception ex)
            throws Exception {
        super.afterCompletion(request, response, handler, ex);
    }

    @Override
    public void afterConcurrentHandlingStarted(HttpServletRequest request,
                                               HttpServletResponse response, Object handler) throws Exception {
        super.afterConcurrentHandlingStarted(request, response, handler);
    }

    @Override
    public void postHandle(HttpServletRequest request,
                           HttpServletResponse response, Object handler,
                           ModelAndView modelAndView) throws Exception {
        super.postHandle(request, response, handler, modelAndView);
    }

}

File: server/src/main/java/org/learning/server/interceptor/LoginInterceptor.kt
package org.learning.server.interceptor

import org.learning.server.common.SessionHelper
import org.learning.server.model.annotation.NoLogin
import org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.HandlerInterceptor
import java.lang.Exception
import javax.security.auth.login.LoginException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class LoginInterceptor : HandlerInterceptor {
    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any?): Boolean {
        if (handler is HandlerMethod && handler.beanType.getAnnotation(NoLogin::class.java) == null) {
            if (handler.beanType == BasicErrorController::class.java) {
                return true
            }

            val user = SessionHelper.of(request).user()

            return when {
                handler.method.isAnnotationPresent(NoLogin::class.java) -> {
                    true
                }
                user == null -> {
                    throw LoginException("当前未登录")
                }
                else -> {
                    true
                }
            }
        }
        return true
    }
}